libraries/SITL/SIM_Sailboat.cpp

```diff
@@ -0,0 +1,180 @@
+{
+    if (fabsf(steering) < 1.0e-6 or fabsf(speed) < 1.0e-6) {
+        return 0;
+    }
+    float d = turn_circle(steering);
+    float c = M_PI * d;
+    float t = c / speed;
+    float rate = 360.0f / t;
+    return rate;
+}
+
+/*
+  return lateral acceleration in m/s/s
+*/
+float Sailboat::calc_lat_accel(float steering_angle, float speed)
+{
+    float yaw_rate = calc_yaw_rate(steering_angle, speed);
+    float accel = radians(yaw_rate) * speed;
+    return accel;
+}
+
+/*
+  update the sailboat simulation by one time step
+ */
+void Sailboat::update(const struct sitl_input &input)
+{
+    // update wind
+    update_wind(input);
+
+    // in sailboats the steering controls the rudder, the throttle controls the main sail position
+    float steering = 2*((input.servos[0]-1000)/1000.0f - 0.5f);
+
+    // calculate mainsail angle from servo output 4, 0 to 90 degrees
+    float mainsail_angle_bf = constrain_float((input.servos[3]-1000)/1000.0f * 90.0f, 0.0f, 90.0f);
+
+    // calculate apparent wind in earth-frame (this is the direction the wind is coming from)
+    Vector3f wind_apparent_ef = wind_ef + velocity_ef;
+    const float wind_apparent_dir_ef = degrees(atan2f(wind_apparent_ef.y, wind_apparent_ef.x));
+    const float wind_apparent_speed = safe_sqrt(sq(wind_apparent_ef.x)+sq(wind_apparent_ef.y));
+
+    // calculate angle-of-attack from wind to mainsail
+    float aoa_deg = MAX(fabsf(wrap_180(wind_apparent_dir_ef - degrees(AP::ahrs().yaw))) - mainsail_angle_bf, 0);
+
+    // calculate Lift force (perpendicular to wind direction) and Drag force (parallel to wind direction)
+    float lift_wf, drag_wf;
+    calc_lift_and_drag(wind_apparent_speed, aoa_deg, lift_wf, drag_wf);
+
+    // rotate lift and drag from wind frame into body frame
+    const float wind_to_veh_rot_angle_deg = wrap_180(180 + wind_apparent_dir_ef - degrees(AP::ahrs().yaw));
+    const float sin_rot_rad = sinf(radians(wind_to_veh_rot_angle_deg));
+    const float cos_rot_rad = cosf(radians(wind_to_veh_rot_angle_deg));
+    const float force_fwd = fabsf((lift_wf * sin_rot_rad)) + (drag_wf * cos_rot_rad);
+
+    // how much time has passed?
+    float delta_time = frame_time_us * 1.0e-6f;
+
+    // speed in m/s in body frame
+    Vector3f velocity_body = dcm.transposed() * velocity_ef;
+
+    // speed along x axis, +ve is forward
+    float speed = velocity_body.x;
+
+    // yaw rate in degrees/s
+    float yaw_rate = calc_yaw_rate(steering, speed);
+
+    gyro = Vector3f(0,0,radians(yaw_rate));
+
+    // update attitude
+    dcm.rotate(gyro * delta_time);
+    dcm.normalize();
+
+    // accel in body frame due acceleration from sail and deceleration from hull friction
+    accel_body = Vector3f((force_fwd * 1.0f) - (velocity_body.x * 0.5f), 0, 0);
+
+    // add in accel due to direction change
+    accel_body.y += radians(yaw_rate) * speed;
+
+    // now in earth frame
+    Vector3f accel_earth = dcm * accel_body;
+    accel_earth += Vector3f(0, 0, GRAVITY_MSS);
```

Comment:
You add a downward acceleration here, then 3 lines later set the entire z field to 0. You can remove this line entirely.

Reply:
txs, removed.